### DISCUSSION TITLE
Cropping wide picture into 3600 separate files.

### DISCUSSION THREAD

**Discussion by W5ALC** [ImageMagick/ImageMagick#5687] [Help] [answered]
I have looked high and low to find a solution to this project i have. 

I have a picture that is roughly 27000x 1080. [The picture in question](https://github.com/W5ALC/ARES/blob/main/Pueblo-Mountains-Big-Natural-Snow.png)
I want to crop a picture that is 1920 x 1080 at around every 7 pixels. I am going to make a sort of moving wallpaper with it just for fun but has turned into a quest to find the solution to this.  Can someone help point me to the best information to acheive my (kind of pointless) goal. Any help would be appreciated.

**Comment by fmw42** [comment]
This is just -sample, but you have to figure out the sample arguments so that it does every 7th pixel in x dimension. You can do that by computing the resize percent for the sample arguments. So for width you would have 100*1920/27000% = 7.11111111%. Height is not reduces so 100%.  So -sample 7.11111x100% should do it.

**Reply by W5ALC** [reply]
i did try it out and while it didnt give the originaly expected results it did create a pretty cool composite of the original 27000 pixels into 1920
![-sample output](https://user-images.githubusercontent.com/7112421/197117905-81523d9b-3c35-4b55-b8f8-9048f7f18e8b.png)

**Comment by snibgo** [comment]
I think the OP wants overlapping crops, and I don't think "-sample" can do that.

What version of IM, on what platform?

To keep the numbers sane, let us suppose your input is 4433x1080 pixels, and you want 360 crops, each 1920x1080. The first image will start at offset x=0, the next at x=7, then x=14, and so on. Hence the overlap (in the x-direction) will be 1920-7 = 1913.

Note that 1920+(360-1)*7 = 4433.

Windows BAT syntax. First, I create an example input:
```
magick -size 4433x1080 xc: samp.png
```
Now we read that, and create 360 crops:
```
magick samp.png -crop 360x1+1913@ out_%%03d.png
```
The 360 outputs are:
```
xc:[0] XC 1920x1080 4433x1080+0+0 16-bit sRGB 0.031u 0:00.032
xc:[1] XC 1920x1080 4433x1080+7+0 16-bit sRGB 0.047u 0:00.048
xc:[2] XC 1920x1080 4433x1080+14+0 16-bit sRGB 0.047u 0:00.059
:
:
xc:[357] XC 1920x1080 4433x1080+2499+0 16-bit sRGB 0.406u 0:05.624
xc:[358] XC 1920x1080 4433x1080+2506+0 16-bit sRGB 0.406u 0:05.636
xc:[359] XC 1920x1080 4433x1080+2513+0 16-bit sRGB 0.406u 0:05.648
```
How much memory (RAM or disk) will we need? We need to store all the input and output pixels. How many pixels are there?

(4433\*1080 + 360\*1920\*1080)/1e9 = 0.751 giga pixels. If each pixel needs 12 bytes, you need 9 GB of RAM or disk.

**Reply by W5ALC** [reply]
i am using Fedora Linux 36 and IM version 6.9.12.64-1.fc36.x86_64

OK i can see that this is definitely more what i am looking for. and as you said with your sample image of 4433x1080 requiring 9 GB of RAM for 360 copies imagine what it would take for 3600 on a 27000x1080. spoiler 7.494 gigapixels.so a bit less than 10 x that amount of RAM. who can afford that kind of setup.

so i think i am going to have to scale down my ridiculous idea to way way less and just deal with it

**Reply by fmw42** [reply]
Sorry, I misunderstood.

**Comment by W5ALC** [comment]
No worries I learned something new so it works out thanks for your help.

**Comment by snibgo** [comment]
> so a bit less than 10 x that amount of RAM. who can afford that kind of setup.

Yes, you need something like 90 GB of memory, but that is RAM and disk. If you don't have enough RAM, IM will use disk, but that will be slow.

**Comment by W5ALC** [comment, marked as answer]
so i case anyone is ever looking for something similar to this i have created a script that does exactly what i was originally looking for. 

Outputting 3600 images moving 7 pixels to the right each one and outputting to a specified directory.

    #!/bin/bash
    set -x
    # Define input and output file paths
    input_file="$1"
    output_directory="/home/nowhereman/Pictures"
    
    # Create output directory if it doesn't exist
    mkdir -p "$output_directory"
    
    # Define image width and height
    image_width=27231
    image_height=1080
    
    # Define split width and height
    split_width=1920
    split_height=1080
    
    # Define x-coordinate offset
    x_offset=0
    
    # Loop through the image and split it into multiple images
    for ((i = 1; i <= 3600; i++)); do
        # Define output file name and path
        output_file="$output_directory/output_image_$i.jpg"
    
        # Split the image using ImageMagick
        convert "$input_file" -crop "${split_width}x${split_height}+${x_offset}+0" "$output_file"
    
        # Increment x-coordinate offset for next split
        x_offset=$((x_offset + 7))
    done

**Comment by snibgo** [comment]
Good stuff, and thanks for posting your script.

However, I strongly suggest modifying your `convert` to do the operations in the order you need: read the input image, crop it, and write it to a file. If you ever upgrade to IM v7 `magick`, the command will need to be in that order.

**Reply by W5ALC** [reply]
Great suggestion i will update it to use  the operations in the correct order.


Also for others use the script to create the xml file to make the background change every 24 seconds itll take the entire day to scan the mountain range.


    #!/bin/bash
    
    # Parse command-line arguments
    if [[ $# -lt 2 || $# -gt 3 ]]; then
      echo "Usage: $0 <images_dir> <xml_file> [--shuffle]"
      exit 1
    fi
    
    images_dir="$1"
    xml_file="$2"
    shuffle=0
    
    if [[ $# -eq 3 && "$3" == "--shuffle" ]]; then
      shuffle=1
    fi
    
    # Validate input arguments
    if [[ ! -d "$images_dir" ]]; then
      echo "Invalid images directory: $images_dir"
      exit 1
    fi
    
    if [[ ! "$xml_file" =~ \.xml$ ]]; then
      echo "Invalid output file: $xml_file"
      exit 1
    fi
    
    # Determine order of images
    if [[ $shuffle -eq 1 ]]; then
      files=($(ls -d -1 $PWD/*.jpg))
    else
      files=($(ls -d -1 $PWD/*.jpg  | sort -V))
    fi
    
    # Write XML file
    cat <<EOF >"$xml_file"
    <background>
      <starttime>
        <year>2022</year>
        <month>08</month>
        <day>04</day>
        <hour>8</hour>
        <minute>00</minute>
        <second>00</second>
      </starttime>
    EOF
    
    for file in "${files[@]}"; do
      cat <<EOF >>"$xml_file"
      <static>
        <duration>24</duration>
        <file>$file</file>
      </static>
    EOF
    done
    
    echo "</background>" >>"$xml_file"

**Comment by snibgo** [comment]
There is a typo in your updated command:
```
convert -crop "$input_file" "${split_width}x${split_height}+${x_offset}+0" "$output_file"
```
But that splits `-crop` from its arguments. It should be:
```
convert "$input_file" -crop "${split_width}x${split_height}+${x_offset}+0" "$output_file"
```

**Comment by W5ALC** [comment]
I ended up making a video of the 3600 files and learned a lot over the last 2.5 years.

[The Mountains from Pueblo](https://youtu.be/xJDlBYkSiPo?si=V16-S30E5HIuyKrp)
